### PR TITLE
optionally delay jobs when retrying them

### PR DESCRIPTION
**Pull request by kwakwaversal**
I am using minion jobs to insert data into a database which needs to survive the database not always being available. When the database isn't available, the jobs are retried straight away without the ability (that I could find) to delay them from being retried too often.

I've added the option `delay` to `job->retry` to delay failed jobs.

I have a requirement to use the delay for several future jobs, so hope this gets considered.

Cheers for writing and supporting Minion et. al.

### CODE REVIEW

**Comment by jberger** [comment]
Is this exposed to the CLI?

File: lib/Minion/Job.pm [hunk 1]
```diff
@@ -225,8 +228,10 @@ Remove C<failed>, C<finished> or C<inactive> job from queue.
 =head2 retry
 
   my $bool = $job->retry;
+  my $bool = $job->retry({delay => 10});
 
-Transition from C<failed> or C<finished> state back to C<inactive>.
+Transition from C<failed> or C<finished> state back to C<inactive>. Optionally
+delay job for several seconds from now.
```
Comment (kraih) [inline lib/Minion/Job.pm]: The documentation is inconsistent with `Minion::enqueue`.

File: lib/Minion/Backend/Pg.pm [hunk 2]
```diff
@@ -128,12 +128,15 @@ sub repair {
 sub reset { shift->pg->db->query('truncate minion_jobs, minion_workers') }
 
 sub retry_job {
```
Comment (kraih) [inline lib/Minion/Backend/Pg.pm]: The documentation for this method has not been updated.

File: lib/Minion/Backend/File.pm [hunk 3]
```diff
@@ -120,13 +120,16 @@ sub repair {
 sub reset { $_[0]->db and delete($_[0]->{db})->clear }
 
 sub retry_job {
```
Comment (kraih) [inline lib/Minion/Backend/File.pm]: The documentation for this method has not been updated.

File: t/pg.t [hunk 4]
```diff
@@ -295,6 +295,15 @@ is $job->id, $id, 'right id';
 ok $job->fail,   'job failed';
 ok $job->remove, 'job has been removed';
 is $job->info,   undef, 'no information';
+$id = $minion->enqueue(add => [6, 9]);
+$job = $worker->dequeue(0);
+like $job->info->{delayed}, qr/^[\d.]+$/, 'has delayed timestamp';
+ok $job->fail, 'job failed';
+ok $job->retry({delay => 10}), 'job retried with delay';
+is $job->info->{retries}, 1, 'job has been retried once';
+ok $job->info->{delayed} > time + 9, 'job delayed';
```
Comment (kraih) [inline t/pg.t]: This test does not leave a lot of room for time differences between servers, i think all other delay tests leave at least 100 seconds.

**Comment by kraih** [comment]
@jberger Good point, doesn't look like it.

**Comment by kwakwaversal** [comment]
Updated the pull request based on your comments.

**Comment by jberger** [comment]
Seems like reasonable functionality. Assuming @kraih is happy with the implementation, I would say :+1:

**Comment by tempire** [comment]
:+1:

**Comment by kraih** [comment]
Thanks, applied.

**Pull request merged**